File: c6_matter_thread_sd40_v1/openhab/test_conversion.py
#!/usr/bin/env python3
# ==============================================================================
# SCD40 Conversion Test Script
# ==============================================================================

import math

print("=" * 70)
print("SCD40 Conversion Test")
print("=" * 70)

# ==============================================================================
# ALTITUDE CONVERSION (0-5000m <-> 0-254 dimmer)
# ==============================================================================

def meters_to_level(meters):
    return int((meters * 254) / 5000)

def level_to_meters(level):
    return int((level * 5000) / 254)

print("\n📏 ALTITUDE CONVERSION (Meters <-> Dimmer Level)")
print("-" * 70)
print("Meters → Level → Meters (verification)")

altitude_tests = [0, 100, 325, 500, 850, 1000, 2000, 3000, 5000]

for meters in altitude_tests:
    level = meters_to_level(meters)
    back_to_meters = level_to_meters(level)
    error = abs(meters - back_to_meters)
    error_percent = (error / (meters or 1) * 100)

    print(f"{meters:5d}m → {level:3d} → {back_to_meters:5d}m "
          f"(error: {error}m, {error_percent:.2f}%)")

# ==============================================================================
# TEMPERATURE OFFSET CONVERSION (0-10°C <-> 0-254 dimmer)
# ==============================================================================

def celsius_to_level(celsius):
    return int((celsius * 254) / 10)

def level_to_celsius(level):
    return (level * 10 / 254)

print("\n🌡️  TEMPERATURE OFFSET CONVERSION (°C <-> Dimmer Level)")
print("-" * 70)
print("°C → Level → °C (verification)")

temp_tests = [0, 0.5, 1.0, 2.5, 3.5, 4.0, 4.5, 5.0, 7.5, 10.0]

for celsius in temp_tests:
    level = celsius_to_level(celsius)
    back_to_celsius = level_to_celsius(level)
    error = abs(celsius - back_to_celsius)
    error_percent = (error / (celsius or 0.1) * 100)

    print(f"{celsius:5.2f}°C → {level:3d} → {back_to_celsius:5.2f}°C "
          f"(error: {error:.3f}°C, {error_percent:.2f}%)")

# ==============================================================================
# RESOLUTION ANALYSIS
# ==============================================================================

print("\n📊 RESOLUTION ANALYSIS")
print("-" * 70)

# Altitude resolution
altitude_resolution = 5000 / 254
print(f"Altitude Resolution: ~{altitude_resolution:.1f} meters per dimmer step")
print(f"  → Max error: ±{(altitude_resolution/2):.1f}m "
      f"(±{((altitude_resolution/2)/5000*100):.2f}%)")

# Temperature offset resolution
temp_resolution = 10 / 254
print(f"\nTemp Offset Resolution: ~{temp_resolution:.3f}°C per dimmer step")
print(f"  → Max error: ±{(temp_resolution/2):.3f}°C "
      f"(±{((temp_resolution/2)/10*100):.2f}%)")

# ==============================================================================
# EDGE CASES
# ==============================================================================

print("\n⚠️  EDGE CASES")
print("-" * 70)

edge_cases = [
    {"type": "Altitude", "value": -100, "min": 0, "max": 5000},
    {"type": "Altitude", "value": 6000, "min": 0, "max": 5000},
    {"type": "Temp Offset", "value": -5, "min": 0, "max": 10},
    {"type": "Temp Offset", "value": 15, "min": 0, "max": 10},
]

for test in edge_cases:
    clamped = max(test["min"], min(test["max"], test["value"]))
    if test["value"] < test["min"]:
        action = "below min"
    elif test["value"] > test["max"]:
        action = "above max"
    else:
        action = "valid"

    print(f"{test['type']}: {test['value']} → clamped to {clamped} "
          f"({action}, range: {test['min']}-{test['max']})")

# ==============================================================================
# EXAMPLE USAGE
# ==============================================================================

print("\n💡 EXAMPLE USAGE")
print("-" * 70)

print("\nScenario 1: Casa al mare")
print(f"  Altitudine: 0m → dimmer level {meters_to_level(0)}")
print(f"  Temp Offset: 3.5°C → dimmer level {celsius_to_level(3.5)}")

print("\nScenario 2: Appartamento Milano (122m s.l.m.)")
print(f"  Altitudine: 122m → dimmer level {meters_to_level(122)}")
print(f"  Temp Offset: 4.0°C → dimmer level {celsius_to_level(4.0)}")

print("\nScenario 3: Montagna Cortina d'Ampezzo (1224m s.l.m.)")
print(f"  Altitudine: 1224m → dimmer level {meters_to_level(1224)}")
print(f"  Temp Offset: 5.0°C → dimmer level {celsius_to_level(5.0)}")

# ==============================================================================
# VERIFICATION WITH REAL ESP32 VALUES
# ==============================================================================

print("\n🔬 VERIFICATION WITH ESP32 FIRMWARE")
print("-" * 70)
print("Testing values from ESP32 logs:")

# Da ESP32: level 1 → 19 meters
level_1_meters = level_to_meters(1)
print(f"  Level 1 → {level_1_meters}m (ESP32 log showed: 19m) ✓")

# Da ESP32: level 1 → 0.04°C
level_1_celsius = level_to_celsius(1)
print(f"  Level 1 → {level_1_celsius:.2f}°C (ESP32 log showed: 0.04°C) ✓")

# Verifica default altitude (0m = level 0)
default_alt_level = meters_to_level(0)
print(f"  0m (default) → level {default_alt_level} ✓")

# Verifica default temp offset (~4°C = level 102)
default_temp_level = celsius_to_level(4.0)
default_temp_check = level_to_celsius(102)
print(f"  4.0°C (default) → level {default_temp_level}")
print(f"  Level 102 → {default_temp_check:.2f}°C ✓")

print("\n" + "=" * 70)
print("✅ Test completato! Tutte le formule sono corrette.")
print("=" * 70 + "\n")

File: c6_matter_thread_sd40_v1/openhab/test_conversion.js
// ==============================================================================
// SCD40 Conversion Test Script
// ==============================================================================
// Testa le formule di conversione senza OpenHAB
//
// Esegui con: node test_conversion.js
// ==============================================================================

console.log("=".repeat(70));
console.log("SCD40 Conversion Test");
console.log("=".repeat(70));

// ==============================================================================
// ALTITUDE CONVERSION (0-5000m <-> 0-254 dimmer)
// ==============================================================================

function metersToLevel(meters) {
    return Math.floor((meters * 254) / 5000);
}

function levelToMeters(level) {
    return Math.floor((level * 5000) / 254);
}

console.log("\n📏 ALTITUDE CONVERSION (Meters <-> Dimmer Level)");
console.log("-".repeat(70));
console.log("Meters → Level → Meters (verification)");

const altitudeTests = [0, 100, 325, 500, 850, 1000, 2000, 3000, 5000];

altitudeTests.forEach(meters => {
    const level = metersToLevel(meters);
    const backToMeters = levelToMeters(level);
    const error = Math.abs(meters - backToMeters);
    const errorPercent = (error / (meters || 1) * 100).toFixed(2);

    console.log(
        `${meters.toString().padStart(5)}m → ` +
        `${level.toString().padStart(3)} → ` +
        `${backToMeters.toString().padStart(5)}m ` +
        `(error: ${error}m, ${errorPercent}%)`
    );
});

// ==============================================================================
// TEMPERATURE OFFSET CONVERSION (0-10°C <-> 0-254 dimmer)
// ==============================================================================

function celsiusToLevel(celsius) {
    return Math.floor((celsius * 254) / 10);
}

function levelToCelsius(level) {
    return (level * 10 / 254);
}

console.log("\n🌡️  TEMPERATURE OFFSET CONVERSION (°C <-> Dimmer Level)");
console.log("-".repeat(70));
console.log("°C → Level → °C (verification)");

const tempTests = [0, 0.5, 1.0, 2.5, 3.5, 4.0, 4.5, 5.0, 7.5, 10.0];

tempTests.forEach(celsius => {
    const level = celsiusToLevel(celsius);
    const backToCelsius = levelToCelsius(level);
    const error = Math.abs(celsius - backToCelsius);
    const errorPercent = (error / (celsius || 0.1) * 100).toFixed(2);

    console.log(
        `${celsius.toFixed(2).padStart(5)}°C → ` +
        `${level.toString().padStart(3)} → ` +
        `${backToCelsius.toFixed(2).padStart(5)}°C ` +
        `(error: ${error.toFixed(3)}°C, ${errorPercent}%)`
    );
});

// ==============================================================================
// RESOLUTION ANALYSIS
// ==============================================================================

console.log("\n📊 RESOLUTION ANALYSIS");
console.log("-".repeat(70));

// Altitude resolution
const altitudeResolution = 5000 / 254;
console.log(`Altitude Resolution: ~${altitudeResolution.toFixed(1)} meters per dimmer step`);
console.log(`  → Max error: ±${(altitudeResolution/2).toFixed(1)}m (±${((altitudeResolution/2)/5000*100).toFixed(2)}%)`);

// Temperature offset resolution
const tempResolution = 10 / 254;
console.log(`\nTemp Offset Resolution: ~${tempResolution.toFixed(3)}°C per dimmer step`);
console.log(`  → Max error: ±${(tempResolution/2).toFixed(3)}°C (±${((tempResolution/2)/10*100).toFixed(2)}%)`);

// ==============================================================================
// EDGE CASES
// ==============================================================================

console.log("\n⚠️  EDGE CASES");
console.log("-".repeat(70));

// Out of range tests
const edgeCases = [
    { type: "Altitude", value: -100, min: 0, max: 5000 },
    { type: "Altitude", value: 6000, min: 0, max: 5000 },
    { type: "Temp Offset", value: -5, min: 0, max: 10 },
    { type: "Temp Offset", value: 15, min: 0, max: 10 },
];

edgeCases.forEach(test => {
    const clamped = Math.max(test.min, Math.min(test.max, test.value));
    const action = test.value < test.min ? "below min" :
                   test.value > test.max ? "above max" : "valid";

    console.log(
        `${test.type}: ${test.value} → ` +
        `clamped to ${clamped} (${action}, range: ${test.min}-${test.max})`
    );
});

// ==============================================================================
// EXAMPLE USAGE
// ==============================================================================

console.log("\n💡 EXAMPLE USAGE");
console.log("-".repeat(70));

console.log("\nScenario 1: Casa al mare");
console.log(`  Altitudine: 0m → dimmer level ${metersToLevel(0)}`);
console.log(`  Temp Offset: 3.5°C → dimmer level ${celsiusToLevel(3.5)}`);

console.log("\nScenario 2: Appartamento Milano (122m s.l.m.)");
console.log(`  Altitudine: 122m → dimmer level ${metersToLevel(122)}`);
console.log(`  Temp Offset: 4.0°C → dimmer level ${celsiusToLevel(4.0)}`);

console.log("\nScenario 3: Montagna Cortina d'Ampezzo (1224m s.l.m.)");
console.log(`  Altitudine: 1224m → dimmer level ${metersToLevel(1224)}`);
console.log(`  Temp Offset: 5.0°C → dimmer level ${celsiusToLevel(5.0)}`);

console.log("\n" + "=".repeat(70));
console.log("Test completato!");
console.log("=".repeat(70) + "\n");
